test/load/test_load_tx_ethbridge_burn.py

```diff
@@ -22,30 +23,103 @@
+    rowan_token_address = burn_rowan_get_erc20_address(ctx)
+    rowan_sc = ctx.get_generic_erc20_sc(rowan_token_address)
+
+    transfer_table = build_transfer_table()
+    amount_per_tx = 1000100101
+
+    _test_load_tx_ethbridge_lock_burn(ctx, amount_per_tx, transfer_table, rowan_sc, isRowan=True)
+
+# test single sif account burn erc20 to multiple ethereum accounts
+def test_single_sif_to_multiple_eth_account_burn_erc20(ctx: test_utils.EnvCtx):
+    # deploy an erc20 contract
+    token_decimals = 18
+    total_amount = 10 ** 28
+    token_data: test_utils.ERC20TokenData = ctx.generate_random_erc20_token_data()
+    erc20_sc = ctx.deploy_new_generic_erc20_token(token_data.name, token_data.symbol, token_decimals)
+    # mint token to operator account
+    ctx.mint_generic_erc20_token(erc20_sc.address, total_amount, ctx.operator) 
+    ctx.advance_blocks()
+
+    transfer_table = build_transfer_table()
+    amount_per_tx = 1000100101
 
-def test_load_tx_ethbridge_burn_short(ctx: test_utils.EnvCtx):
-    _test_load_tx_ethbridge_burn(ctx, 123456 * eth.GWEI, [[2, 2], [2, 2]])
+    _test_load_tx_ethbridge_lock_burn(ctx, amount_per_tx, transfer_table, erc20_sc)
 
+# test single sif account burn ceth to multiple ethereum accounts
+def test_single_sif_to_multiple_eth_account_burn_eth(ctx: test_utils.EnvCtx):
+    transfer_table = build_transfer_table()
+    amount_per_tx = 1000100101
+    eth_sc = ctx.get_generic_erc20_sc(NULL_ADDRESS)
 
-def test_load_tx_ethbridge_burn(ctx: test_utils.EnvCtx):
+    _test_load_tx_ethbridge_lock_burn(ctx, amount_per_tx, transfer_table, eth_sc)
+
+# short test to verify eth burn works
+def test_load_tx_ethbridge_burn_eth_short(ctx: test_utils.EnvCtx):
+    transfer_table = [[2, 2], [2, 2]]
+    amount_per_tx = 1000100101
+    eth_sc = ctx.get_generic_erc20_sc(NULL_ADDRESS)
+    _test_load_tx_ethbridge_lock_burn(ctx, amount_per_tx, transfer_table, eth_sc)
+
+# test multiple sif accounts burn ceth to multiple ethereum accounts
+def test_load_tx_ethbridge_burn_eth(ctx: test_utils.EnvCtx):   
     # Matrix of transactions that we want to send. A row (list) in the table corresponds to a sif account sending
     # transactions to eth accounts. The numbers are transaction counts, where each transaction is for amount_per_tx.
     # Each sif account uses a dedicated send thread.
     transfer_table = [
-        [100, 100, 100],
-        [100, 100, 100],
-        [100, 100, 100],
-        [10, 20, 30],
-    ]
-
-    amount_per_tx = 123456 * eth.GWEI
-
-    _test_load_tx_ethbridge_burn(ctx, amount_per_tx, transfer_table)
-
+            [100, 100, 100],
+            [100, 100, 100],
+            [100, 100, 100],
+            [10, 20, 30],
+        ]
+    amount_per_tx = 1000100101
+    eth_sc = ctx.get_generic_erc20_sc(NULL_ADDRESS)
+    _test_load_tx_ethbridge_lock_burn(ctx, amount_per_tx, transfer_table, eth_sc)
+
+def _test_load_tx_ethbridge_lock_burn(ctx: test_utils.EnvCtx, amount_per_tx: int, 
```

Comment:
It would be very helpful for me if I could use the code in this test to understand better the code that is being tested. I tried to read this procedure, taking notes in a piece of paper, but two hours later I gave up. It's too complicated.

Reply:
This test seems to have multiple stages, which should be separated in their own named procedures:

- Validation of inputs.
- Setup of test, which itself can be split into multiple stages:
  - Creation and funding of accounts on the Sifchain side.
  - Creation and funding of accounts on the Ethereum side.
- One or more of these actions:
  - Execution of test transactions.
  - Validation of results of test transactions.

It is much easier to understand that way, instead of having one long procedure where all these stages are mushed together.

A comment like this where the `transfer_table` parameter of the function is declared would have helped:

> Rows are Sifchain accounts, Columns are Ethereum accounts, numbers are the count of transactions to perform for each pair.

I thought for a long while that those numbers represented amounts of a token, and nothing made much sense.